src/cljs/harja/tiedot/hallinta/urakoiden_lyhytnimet.cljs

```diff
@@ -0,0 +1,91 @@
+(ns harja.tiedot.hallinta.urakoiden-lyhytnimet
+  (:require [reagent.core :refer [atom]]
+            [tuck.core :as tuck]
+            [taoensso.timbre :as log]
+            [harja.tyokalut.tuck :as tuck-apurit]
+            [harja.ui.viesti :as viesti]))
+
+(defonce tila (atom {:valittu-urakkatyyppi {:nimi "Kaikki" :arvo :kaikki}
+                     :vain-puuttuvat false
+                     :urakan-tila :kaikki}))
+
+(defrecord HaeUrakoidenLyhytnimet [valinnat])
+(defrecord HaeUrakoidenLyhytnimetOnnistui [vastaus])
+(defrecord HaeUrakoidenLyhytnimetEpaonnistui [vastaus])
+(defrecord PaivitaUrakoidenLyhytnimet [urakat])
+(defrecord PaivitaUrakoidenLyhytnimetOnnistui [vastaus])
+(defrecord PaivitaUrakoidenLyhytnimetEpaonnistui [vastaus])
+(defrecord PaivitaValittuUrakkaTyyppi [valittu-urakkatyyppi])
+(defrecord PaivitaVainPuuttuvat [vain-puuttuvat])
+(defrecord PaivitaUrakanTila [urakan-tila])
+
+(defn- hae-urakoiden-nimet [parametrit]
+  (tuck-apurit/post! :hae-urakoiden-nimet parametrit
+    {:onnistui ->HaeUrakoidenLyhytnimetOnnistui
+     :epaonnistui ->HaeUrakoidenLyhytnimetEpaonnistui}))
+
+(defn- parsi-urakkatyyppi [app]
+  (:arvo (:valittu-urakkatyyppi app)))
+
+(defn- hakuparametrit [app]
+   {:urakkatyyppi (parsi-urakkatyyppi app)
+    :vain-puuttuvat (:vain-puuttuvat app)
+    :urakan-tila (:urakan-tila app)})
+
+(extend-protocol tuck/Event
+
+  HaeUrakoidenLyhytnimet
+  (process-event [_ app]
+    (do
+      (hae-urakoiden-nimet (hakuparametrit app))
+      (assoc app :hae-urakoiden-lyhytnimet-kesken? true)))
+
+  HaeUrakoidenLyhytnimetOnnistui
+  (process-event [{vastaus :vastaus} app]
+    (assoc app
+      :urakoiden-lyhytnimet vastaus
+      :hae-urakoiden-lyhytnimet-kesken? false))
+
+  HaeUrakoidenLyhytnimetEpaonnistui
+  (process-event [{vastaus :vastaus} app]
+    (log/error "Urakoiden nimien haku epäonnistui. Virhe: " vastaus)
+    (viesti/nayta-toast! "Urakoiden nimien haku epäonnistui." :varoitus)
+    (assoc app :hae-urakoiden-lyhytnimet-kesken? false))
+
+  PaivitaUrakoidenLyhytnimet
+  (process-event [{urakat :urakat} app]
+    (do
+      (tuck-apurit/post! :tallenna-urakoiden-lyhytnimet (assoc urakat :haku-parametrit (hakuparametrit app))
```

Comment:
Tehdään mielummin näin, vähän selkeämpi:

```
(tuck-apurit/post! :tallenna-urakoiden-lyhytnimet 
        {:tiedot urakat :haku-parametrit (hakuparametrit app)}
        {:onnistui ->PaivitaUrakoidenLyhytnimetOnnistui
         :epaonnistui ->PaivitaUrakoidenLyhytnimetEpaonnistui})```

Reply:
nyt on tehty noin, mutta tuosta tuli ylimääräinen avain, arvo taso